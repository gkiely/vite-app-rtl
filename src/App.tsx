import { useEffect, useState } from 'react';
import logo from './logo.svg';
import * as styles from './App.css';
import Button from './components/Button';
import { Post, postsSchema } from '../server/worker';

const fetchPosts = async (s: string): Promise<Post[]> => {
  try {
    const response = await fetch(s);
    const data = await response.json();
    return postsSchema.parse(data);
  } catch {
    return [];
  }
};

function App() {
  const [count, setCount] = useState(0);
  const [posts, setPosts] = useState<Post[]>([]);
  useEffect(() => {
    fetchPosts('/api/posts').then(setPosts);
  }, []);

  return (
    <div className={styles.app}>
      <header className={styles.header}>
        <img src={logo} className={styles.logo} alt="logo" />
        <p>Hello Vite + React!</p>
        <p>
          <Button type="button" onClick={() => setCount(count => count + 1)}>
            count is: {count}
          </Button>
        </p>
        <p>
          Update <code>App.tsx</code> and save to test HMR updates.
        </p>
        <p>
          <a
            className={styles.link}
            href="https://reactjs.org"
            target="_blank"
            rel="noopener noreferrer"
          >
            Learn React
          </a>
          {' | '}
          <a
            className={styles.link}
            href="https://vitejs.dev/guide/features.html"
            target="_blank"
            rel="noopener noreferrer"
          >
            Vite Docs
          </a>
        </p>
      </header>
      {posts.map(post => (
        <div key={post.id}>
          <h1>{post.title}</h1>
        </div>
      ))}
    </div>
  );
}

export default App;
