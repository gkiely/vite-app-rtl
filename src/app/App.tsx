import { createElement, useState } from 'react';
import { useLocation } from 'react-router-dom';
import * as styles from './App.css';
import useRoute from 'utils/useRoute';
import { Path, renderers } from 'routes/routes';
import { store } from 'routes/server';
import * as Components from './components';
import RouteContext from './RouteContext';
import { assertType } from 'utils';
import { Intersect } from 'utils/types';
import { RouteConfig } from 'utils/routing';

/* c8 ignore start */
const Route = (_: { key: string }) => {
  const location = useLocation();
  assertType<Path>(location.pathname);

  const render = renderers[location.pathname];
  const [route, setRoute] = useState<RouteConfig>(render(store));
  const send = useRoute(setRoute);

  return (
    <RouteContext send={send}>
      <div className={styles.app}>
        {route.map((props) => {
          const Component = Components[props.component];
          if (props.id) {
            assertType<{ key: string }>(props);
            props.key = props.id;
          }

          // Convert union to intersection type for dynamic components
          assertType<Intersect<typeof props>>(props);
          assertType<Intersect<typeof Component>>(Component);
          return createElement(Component, props);
        })}
      </div>
    </RouteContext>
  );
};
/* c8 ignore stop */

function App() {
  const location = useLocation();

  // Render the route
  return <Route key={location.pathname} />;
}

export default App;
