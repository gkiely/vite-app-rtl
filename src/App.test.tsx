import { fireEvent, render, screen } from '@testing-library/react';

import App from './App';

describe('App', () => {
  it('should render', () => {
    render(<App />);
    expect(screen.getByText('Hello Vite + React!')).toBeInTheDocument();
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(screen.getByText('count is: 1')).toBeInTheDocument();
  });
});
