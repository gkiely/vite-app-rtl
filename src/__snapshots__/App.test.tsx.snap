// Vitest Snapshot v1

exports[`App > should render 1`] = `
<DocumentFragment>
  <div
    class="App__4h283f1"
  >
    <header
      class="App__4h283f3"
    >
      <img
        alt="logo"
        class="App__4h283f2"
        src="/src/logo.svg"
      />
      <p>
        Hello Vite + React!
      </p>
      <p>
        <button
          class="Button__1bfezsx0"
          type="button"
        >
          count is: 0
        </button>
      </p>
      <p>
        Update 
        <code>
          App.tsx
        </code>
         and save to test HMR updates.
      </p>
      <p>
        <a
          class="App__4h283f4"
          href="https://reactjs.org"
          rel="noopener noreferrer"
          target="_blank"
        >
          Learn React
        </a>
         | 
        <a
          class="App__4h283f4"
          href="https://vitejs.dev/guide/features.html"
          rel="noopener noreferrer"
          target="_blank"
        >
          Vite Docs
        </a>
      </p>
    </header>
  </div>
</DocumentFragment>
`;
